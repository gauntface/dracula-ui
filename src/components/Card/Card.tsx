import React from 'react'
import { Box, BoxProps } from '@/components/Box/Box'
import { borderColors } from '@/base/colors'
import cx from 'classnames'

export const cardOrientations = {
  normal: '',
  portrait: 'drac-card-portrait'
}
export type CardOrientationProps = keyof typeof cardOrientations

export const cardVariants = {
  normal: '',
  subtle: 'drac-card-subtle'
}
export type CardVariantProps = keyof typeof cardVariants

export interface CardProps extends BoxProps {
  orientation?: CardOrientationProps
  variant?: CardVariantProps
}

export const Card: React.FC<CardProps> = (props: CardProps) => {
  const classNames = cx(
    'drac-card',
    props.className,
    props.orientation && cardOrientations[props.orientation],
    props.variant && cardVariants[props.variant],
    // apply border color based on theme color
    props.color && props.variant && borderColors[props.color]
  )

  return (
    <Box {...props} className={classNames}>
      {props.children}
    </Box>
  )
}

Card.displayName = 'Card'
