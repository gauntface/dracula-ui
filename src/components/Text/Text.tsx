import cx from 'classnames/dedupe'
import { mapValues } from 'lodash'
import React, { HTMLAttributes } from 'react'
import { colors } from '../../base/colors'
import {
  marginMixin,
  MarginMixin,
  paddingMixin,
  PaddingMixin
} from '../../base/spacing'

export const textSizes = {
  large: 'drac-text-lg',
  medium: 'drac-text',
  small: 'drac-text-sm',
  xsmall: 'drac-text-xs'
}

export const textWeights = {
  normal: 'drac-text',
  semibold: 'drac-text-semibold',
  bold: 'drac-text-bold'
}

export const textColors = mapValues(colors, (className) => {
  return className.replace('-bg-', '-text-')
})

/** Text Props */
export interface TextProps
  extends HTMLAttributes<HTMLSpanElement>,
    PaddingMixin,
    MarginMixin {
  /**
   * Controls the size of the text based on pre-configured Dracula UI sizes.
   * Options: `xsmall`, `small`, `medium`, `large`
   */
  size?: keyof typeof textSizes

  /**
   * Controsl the weight of the text.
   * Options: `normal`, `semibold`, `bold`.
   */
  weight?: keyof typeof textWeights

  /**
   * Controls the color of the text
   */
  color?: keyof typeof textColors

  as?: 'a' | 'span' | 'p'
}

/**
 * Text is the base component for any sort of text.
 *
 * Consumers of this component can control, the color, size,
 * weight, and spacing of Text.
 *
 * Use this component for generic, and non-hierarchical text that is
 * to be displayed on a page, or as part of other complex components
 * or UI patterns
 */
export const Text = (props: TextProps) => {
  const finalProps = {
    ...props,

    className: cx(
      `drac-text`,
      textSizes[props.size ?? 'medium'],
      textWeights[props.weight ?? 'normal'],
      textColors[props.color ?? 'white'],
      ...paddingMixin(props),
      ...marginMixin(props)
    )
  }

  return React.createElement(
    finalProps.as ?? 'span',
    finalProps,
    props.children
  )
}

Text.displayName = 'Text'
