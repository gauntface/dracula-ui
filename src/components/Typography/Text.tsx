import React, { HTMLAttributes } from 'react'
import { mapValues } from 'lodash'
import cx from 'classnames'
import { spacingClasses, SpacingPropType } from '@/base/spacing'
import { textColors } from '@/base/colors'

export const TextSizes = {
  large: 'drac-text-lg',
  medium: 'drac-text-md',
  normal: 'drac-text',
  small: 'drac-text-sm'
}

export const TextWeights = {
  normal: 'drac-text',
  semibold: 'drac-text-semibold',
  bold: 'drac-text-bold'
}

export const TextColors = mapValues(textColors, (className) => {
  return className.replace('-bg-', '-text-')
})

export interface TextProps extends HTMLAttributes<HTMLSpanElement> {
  size?: keyof typeof TextSizes
  weight?: keyof typeof TextWeights
  color?: keyof typeof TextColors
  spacing?: SpacingPropType
}

export const Text = (props: TextProps) => {
  const finalProps = {
    ...props,

    className: cx(
      `drac-text`,
      TextSizes[props.size ?? 'normal'],
      TextWeights[props.weight ?? 'normal'],
      TextColors[props.color ?? 'black'],
      spacingClasses(props.spacing)
    )
  }

  return <span {...finalProps}>{props.children}</span>
}

export const Paragraph: React.FC<TextProps> = (props: TextProps) => {
  const finalProps = {
    ...props,

    className: cx(
      `drac-text`,
      TextSizes[props.size ?? 'normal'],
      TextWeights[props.weight ?? 'normal'],
      TextColors[props.color ?? 'black'],
      spacingClasses(props.spacing ?? 'smallY')
    )
  }

  return <p {...finalProps}>{props.children}</p>
}

Text.displayName = 'Text'
