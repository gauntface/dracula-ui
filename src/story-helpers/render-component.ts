import testRenderer from 'react-test-renderer'

export type ComponentExample = {
  code: ReturnType<testRenderer.ReactTestRenderer['toJSON']>
  docs: string
  title: string
}

export type SnapshotBuilder = {
  comp: () => React.ReactElement
  docs: string
  title: string
}

export function buildSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): SnapshotBuilder {
  return {
    title,
    comp,
    docs: docs.trim()
  }
}

export function renderSnapshot(
  title: string,
  comp: React.ReactElement,
  docs: string = ''
): ComponentExample {
  const rendered = testRenderer.create(comp)
  const code = rendered.toJSON()
  return {
    title,
    code,
    docs
  }
}
