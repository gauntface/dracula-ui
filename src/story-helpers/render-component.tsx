import React from 'react'
import Enzyme, { mount,  shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import prettier from 'prettier'
import toJson from "enzyme-to-json";

Enzyme.configure({ adapter: new Adapter() })

export type ComponentExample = {
  html: string
  docs: string
  title: string
  react?: any
}

export type SnapshotBuilder = {
  comp: () => React.ReactElement
  docs: string
  title: string
}

export function buildSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): SnapshotBuilder {
  return {
    title,
    comp,
    docs: docs.trim()
  }
}

export function renderSnapshot(
  title: string,
  comp: React.ReactElement,
  docs: string = ''
): ComponentExample {

  
  const html = mount(comp).html()
  const react = toJson(shallow(<div>{comp}</div>))

  const prettyHTML = prettier.format(html, {
    parser: 'html'
  })

  return {
    title,
    html: prettyHTML,
    react,
    docs
  }
}
