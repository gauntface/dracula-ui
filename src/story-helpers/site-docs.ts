import {
  renderSnapshot,
  SnapshotBuilder
} from '@/story-helpers/render-component'
import React from 'react'

// import * as serializer from 'jest-snapshot-serializer-raw'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  describe(`Site: ${componentClass.displayName ?? componentClass.name}`, () => {
    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp(),
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
    })

    const variations = documentation.variations()

    variations.forEach(variation => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp(),
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()
      })
    })
  })
}
