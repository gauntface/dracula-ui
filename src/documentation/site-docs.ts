import {
  ComponentExample,
  renderSnapshot,
  SnapshotBuilder
} from '../story-helpers/render-component'
import React from 'react'
import fs from 'fs-extra'
import { pretty } from '../story-helpers/pretty'
import { getDocGen } from '../documentation/docgen/doc-generator'
import { toDSP } from '../documentation/dsp/component-generator'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  const name = componentClass.displayName ?? componentClass.name

  describe(`Site: ${name}`, () => {
    let examples: Record<string, ComponentExample> = {}

    afterAll(async () => {
      const docGen = getDocGen(name)
      const dsp = toDSP(name, examples, docGen)
      const dspPath = `${process.cwd()}/dsp/data/components/${name}.json`
      await fs.writeFile(dspPath, pretty(JSON.stringify(dsp), 'json'))
    })

    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp,
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
      examples = {
        ...examples,
        basic: snapshot
      }
    })

    const variations = documentation.variations()

    variations.forEach((variation) => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp,
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()

        examples = {
          ...examples,
          [variation.title]: snapshot
        }
      })
    })
  })
}
