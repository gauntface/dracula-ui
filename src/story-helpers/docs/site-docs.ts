import {
  ComponentExample,
  renderSnapshot,
  SnapshotBuilder
} from '@/story-helpers/render-component'
import React from 'react'
import fs from 'fs'
import { pretty } from '@/story-helpers/pretty'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  const name = componentClass.displayName ?? componentClass.name

  describe(`Site: ${name}`, () => {
    let examples: Record<string, ComponentExample> = {}

    afterAll(() => {
      const path = `${process.cwd()}/examples/${name}.json`
      fs.writeFileSync(path, pretty(JSON.stringify(examples), 'json'))
    })

    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp,
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
      examples = {
        ...examples,
        basic: snapshot
      }
    })

    const variations = documentation.variations()

    variations.forEach((variation) => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp,
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()

        examples = {
          ...examples,
          [variation.title]: snapshot
        }
      })
    })
  })
}
