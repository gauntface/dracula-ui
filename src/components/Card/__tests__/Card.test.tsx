import React from 'react'
import { docs } from '@/documentation/site-docs'
import { snapshot } from '@/documentation/render-component'
import { Text } from '@/components/Text/Text'
import { Card } from '@/components/Card/Card'

docs(Card, {
  basic() {
    return snapshot(
      'Usage',
      Usage,
      'A Card is analogous to an HTML div and can contain any other component. Use Cards when you want to add hierarchy or highlight parts of your content.'
    )
  },
  variations: () => [
    snapshot(
      'Colors',
      Colors,
      'Cards can be customized to use any of the Dracula UI theme colors, including all solid, gradient, and animated colors.'
    ),

    snapshot(
      'Variants',
      Variants,
      'Use variant="subtle" to create discrete or secondary cards'
    ),

    snapshot(
      'Spacing',
      Spacing,
      'Cards come with pre defined spacing primitives for padding and margin that allow for customizing the spacing inside and around Cards.'
    )
  ]
})

function Usage() {
  return (
    <Card color="purple" p="sm">
      <Text color="black">Dracula</Text>
    </Card>
  )
}

function Colors() {
  return (
    <Card color="pinkPurple" p="sm">
      <Text color="black">Pink Purple</Text>
    </Card>
  )
}

function Variants() {
  return (
    <div>
      <Card variant="subtle" color="pink" p="md" m="md">
        <Text color="pink">Subtle Card</Text>
      </Card>
    </div>
  )
}

function Spacing() {
  return (
    <div>
      <Card color="yellowPink" p="xs">
        <Text color="black">xs padding</Text>
      </Card>

      <Card color="purpleCyan" p="sm">
        <Text color="black">sm padding</Text>
      </Card>

      <Card color="pinkPurple" p="md">
        <Text color="black">medium padding</Text>
      </Card>

      <Card color="yellowPink" m="xs">
        <Text color="black">xs margin</Text>
      </Card>

      <Card color="purpleCyan" m="sm">
        <Text color="black">sm margin</Text>
      </Card>

      <Card color="pinkPurple" m="md">
        <Text color="black">medium margin</Text>
      </Card>
    </div>
  )
}
