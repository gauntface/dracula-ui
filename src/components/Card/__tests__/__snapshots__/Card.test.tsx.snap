// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site: Card Basic Usage 1`] = `
Object {
  "docs": "A Card is analogous to an HTML div and can contain any other component. Use Cards when you want to add hierarchy or highlight parts of your content.",
  "html": "<div color=\\"purple\\" p=\\"sm\\" class=\\"drac-box drac-card drac-bg-purple drac-p-sm\\">
  <span class=\\"drac-text drac-line-height drac-text-white\\">Dracula</span>
</div>
",
  "react": "<Card color=\\"purple\\" p=\\"sm\\">
  <Text>Dracula</Text>
</Card>;
",
  "title": "Usage",
}
`;

exports[`Site: Card Colors 1`] = `
Object {
  "docs": "Cards can be customized to use any of the Dracula UI theme colors, including all solid, gradient, and animated colors.",
  "html": "<div
  color=\\"pinkPurple\\"
  p=\\"sm\\"
  class=\\"drac-box drac-card drac-bg-pink-purple drac-p-sm\\"
>
  <span class=\\"drac-text drac-line-height drac-text-white\\">Pink Purple</span>
</div>
",
  "react": "<Card color=\\"pinkPurple\\" p=\\"sm\\">
  <Text>Pink Purple</Text>
</Card>;
",
  "title": "Colors",
}
`;

exports[`Site: Card Spacing 1`] = `
Object {
  "docs": "Cards come with pre defined spacing primitives for padding and margin that allow for customizing the spacing inside and around Cardes.",
  "html": "<div>
  <div
    color=\\"yellowPink\\"
    p=\\"xs\\"
    class=\\"drac-box drac-card drac-bg-yellow-pink drac-p-xs\\"
  >
    <span color=\\"black\\" class=\\"drac-text drac-line-height drac-text-black\\"
      >xs padding</span
    >
  </div>
  <div
    color=\\"purpleCyan\\"
    p=\\"sm\\"
    class=\\"drac-box drac-card drac-bg-purple-cyan drac-p-sm\\"
  >
    <span color=\\"black\\" class=\\"drac-text drac-line-height drac-text-black\\"
      >sm padding</span
    >
  </div>
  <div
    color=\\"pinkPurple\\"
    p=\\"md\\"
    class=\\"drac-box drac-card drac-bg-pink-purple drac-p-md\\"
  >
    <span color=\\"black\\" class=\\"drac-text drac-line-height drac-text-black\\"
      >medium padding</span
    >
  </div>
  <div
    color=\\"yellowPink\\"
    m=\\"xs\\"
    class=\\"drac-box drac-card drac-bg-yellow-pink drac-m-xs\\"
  >
    <span color=\\"black\\" class=\\"drac-text drac-line-height drac-text-black\\"
      >xs margin</span
    >
  </div>
  <div
    color=\\"purpleCyan\\"
    m=\\"sm\\"
    class=\\"drac-box drac-card drac-bg-purple-cyan drac-m-sm\\"
  >
    <span color=\\"black\\" class=\\"drac-text drac-line-height drac-text-black\\"
      >sm margin</span
    >
  </div>
  <div
    color=\\"pinkPurple\\"
    m=\\"md\\"
    class=\\"drac-box drac-card drac-bg-pink-purple drac-m-md\\"
  >
    <span color=\\"black\\" class=\\"drac-text drac-line-height drac-text-black\\"
      >medium margin</span
    >
  </div>
</div>
",
  "react": "<div>
  <Card color=\\"yellowPink\\" p=\\"xs\\">
    <Text color=\\"black\\">xs padding</Text>
  </Card>
  <Card color=\\"purpleCyan\\" p=\\"sm\\">
    <Text color=\\"black\\">sm padding</Text>
  </Card>
  <Card color=\\"pinkPurple\\" p=\\"md\\">
    <Text color=\\"black\\">medium padding</Text>
  </Card>
  <Card color=\\"yellowPink\\" m=\\"xs\\">
    <Text color=\\"black\\">xs margin</Text>
  </Card>
  <Card color=\\"purpleCyan\\" m=\\"sm\\">
    <Text color=\\"black\\">sm margin</Text>
  </Card>
  <Card color=\\"pinkPurple\\" m=\\"md\\">
    <Text color=\\"black\\">medium margin</Text>
  </Card>
</div>;
",
  "title": "Spacing",
}
`;

exports[`Site: Card Variants: subtle 1`] = `
Object {
  "docs": "Use variant=\\"subtle\\" to create discrete or secondary cards",
  "html": "<div>
  <div
    variant=\\"subtle\\"
    color=\\"pink\\"
    p=\\"md\\"
    m=\\"md\\"
    class=\\"drac-box drac-card drac-card-subtle drac-border-pink drac-bg-pink drac-p-md drac-m-md\\"
  >
    <span color=\\"pink\\" class=\\"drac-text drac-line-height drac-text-pink\\"
      >Subtle Card</span
    >
  </div>
</div>
",
  "react": "<div>
  <Card variant=\\"subtle\\" color=\\"pink\\" p=\\"md\\" m=\\"md\\">
    <Text color=\\"pink\\">Subtle Card</Text>
  </Card>
</div>;
",
  "title": "Variants: subtle",
}
`;
